common/src/main/java/com/microsoft/identity/common/internal/ui/webview/challengehandlers/NfcSmartcardCertBasedAuthChallengeHandler.java

```diff
@@ -0,0 +1,170 @@
+    /**
+     * Creates new instance of NfcSmartcardCertBasedAuthChallengeHandler.
+     * @param activity current host activity.
+     * @param nfcSmartcardCertBasedAuthManager AbstractNfcSmartcardCertBasedAuthManager instance.
+     * @param dialogHolder DialogHolder instance.
+     * @param telemetryHelper CertBasedAuthTelemetryHelder instance.
+     */
+    public NfcSmartcardCertBasedAuthChallengeHandler(@NonNull final Activity activity,
+                                                     @NonNull final AbstractNfcSmartcardCertBasedAuthManager nfcSmartcardCertBasedAuthManager,
+                                                     @NonNull final DialogHolder dialogHolder,
+                                                     @NonNull final CertBasedAuthTelemetryHelper telemetryHelper) {
+        super(activity, nfcSmartcardCertBasedAuthManager, dialogHolder, telemetryHelper, NfcSmartcardCertBasedAuthChallengeHandler.class.getSimpleName());
+    }
+
+    /**
+     * Additional logic to run upon before the end of processChallenge.
+     */
+    @Override
+    protected void onGetSessionFinished() {
+        mSmartcardCertBasedAuthManager.stopDiscovery(mActivity);
+    }
+
+    /**
+     * Upon a positive button click in the smartcard PIN dialog, verify the provided PIN and handle the results.
+     * @param certDetails ICertDetails of the selected certificate from the SmartcardCertPickerDialog.
+     * @param request ClientCertRequest received from AzureActiveDirectoryWebViewClient.onReceivedClientCertRequest.
+     * @return A PositiveButtonListener to be set for a SmartcardPinDialog.
+     */
+    @Override
+    protected SmartcardPinDialog.PositiveButtonListener getSmartcardPinDialogPositiveButtonListener(@NonNull final ICertDetails certDetails,
+                                                                                                    @NonNull final ClientCertRequest request) {
+        final String methodTag = TAG + ":getSmartcardPinDialogPositiveButtonListener";
+
+        return new SmartcardPinDialog.PositiveButtonListener() {
+            @RequiresApi(api = Build.VERSION_CODES.LOLLIPOP)
+            @Override
+            public void onClick(@NonNull final char[] pin) {
+                //For NFC, we need another dialog prompting the user to hold the smartcard to the phone again.
+                mDialogHolder.showSmartcardNfcPromptDialog(new SmartcardNfcPromptDialog.CancelCbaCallback() {
+                    @Override
+                    public void onCancel() {
+                        mDialogHolder.dismissDialog();
+                        mTelemetryHelper.setResultFailure(USER_CANCEL_MESSAGE);
+                        request.cancel();
+                    }
+                });
+                mSmartcardCertBasedAuthManager.setConnectionCallback(new AbstractSmartcardCertBasedAuthManager.IConnectionCallback() {
+                    @Override
+                    public void onCreateConnection() {
```

Comment:
(maybe we could turn onCreateConnection setter to clear itself once the callback is fired).

Reply:
I think this could be done for NFC, but not needed for USB. So a method to clear itself would only be applied to NFC classes and not base ones. That's something I'll look into.